### ISSUE TITLE
Misleading error message for code with associated types and lifetimes

### ISSUE THREAD

**Issue by smmalis37** [rust-lang/rust#74736]
The following code

```rust
struct S {}

trait T<'a>{
    type A;
}

impl T<'_> for S {
    type A = i32;
}

fn foo<'a>(x: impl Fn(<S as T<'_>>::A) -> <S as T<'a>>::A) {
    x(2i32);
}

fn main() {
    foo(|x| x);
}
```

currently produces the following error:

```
error[E0308]: mismatched types
  --> src/main.rs:12:7
   |
12 |     x(2i32);
   |       ^^^^ expected associated type, found `i32`
   |
   = note: expected associated type `<S as T<'_>>::A`
                         found type `i32`
```

This is confusing because at a glance it would seem like the error is nonsensical. The associated type **is** i32, so how could they mismatch? I believe the real problem is that because there are no input references, no lifetime elision is taking place, meaning there is no '_ in scope for that function. The error message should probably be changed to something that actually mentions lifetimes (if I'm correct in my understanding of the real issue).

**Comment by smmalis37** [comment]
Closing in favor of #76956
